Updated testing infrastructure

lucass-carneiro opened the pull request:
I have made the following changes

1 (AsterX) Added a new unit testing infrastructure
2 (AsterX) Added/updated a few missing header guards

Right now, the unit tests are not as extensive as in the original code. This PR is intended to be a proposal on how unit tests would be implemented/integrated.

The basic idea is to have a special source folder to contain individual tests and run them all at once in the cactus scheduled function. Using individual files for each test is interesting if testing code gets particularly large for some functionality, but this can be refactored later.

I've also added some color to pass/fail messages. That should make them easier to spot

![Screenshot_20230428_003851](https://user-images.githubusercontent.com/17052910/235049687-87d3b928-76c6-45ad-896e-0dec084eb650.png)

lucass-carneiro commented:
An important detail I forgot to mention: All tests construct objects with random input in them. This is probably not appropriate to all kinds of tests, but those where we check mathematical properties and correctness, that can be useful. The random seed is fixed to generate reproducible results

lennoggi commented on AsterX/src/test.cxx:
```diff
@@ -1,188 +1,17 @@
 
-    CCTK_VINFO("Test vec of smats succeeded");
-  }
+  test_corss_product();
```
*cross

lennoggi commented on AsterX/src/test.hxx:
```diff
@@ -0,0 +1,46 @@
+void test_contraction_smat_upvec();
+void test_contraction_upvec_downvec();
+
+void test_corss_product();
```
*cross

stevenrbrandt commented:
I think we need a general way to do unit tests in Cactus. I'm wondering whether some of the methods in this PR should move to the flesh and whether some conventions about test names and locations can be established. @rhaas80, please comment.

rhaas80 commented:
BTW this discussion may be best done in a ticket in the ET ticketing system so that any resolution can be captured and archived

rhaas80 commented:
Looking at the screenshot. I suggest to avoid using colors (or any escape sequences) since they will not look ok if one looks at the log using less (unless one passes options) or vim/view/emacs/VSCode.

lucass-carneiro commented:
On having a unified way of doing unit tests:

That would be ideal, but hard to solve, because like it was said, this requires consensus, and is a `ET` issue. I have no intention of solving that here. The intention is to lay a foundation to do testing *within* `AsterX`.

My main main goal is to test, e.g., reconstruction, with the code as it is right now, so I created a sort of framework for doing that and making things easier for now. One day when the generalized framework exists, this internal tester could be scrapped/rewritten to use the generalized on in `Cactus`.

As for changes to the original PR:

* I have added support for repeating the randomized tests a number of times, randomizing new values each time.
* I have changed the PR into a "Draft PR". Erik told me that this *should* trigger the CI tests to run, addressing one of Jay's main concerns. Once the CI tests run, it can be converted back to a regular PR and merged.

lucass-carneiro commented:
Actually, I think that in order to run the CI tests, `pull_request` needs to be added to the workflow file. See [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request)
